include/deal.II/base/table_indices.h

```diff
@@ -43,11 +45,71 @@ class TableIndices
 public:
 
   /**
-   * Default constructor. It sets all indices to zero.
+   * Default constructor. This constructor sets all indices to zero.
    */
   TableIndices();
 
   /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<1> and initializes the single
+   * index with @p index0.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from one.
+   */
+  explicit TableIndices (const unsigned int index0);
+
+  /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<2> and initializes the
+   * indices stored by this object by the given arguments.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from two.
+   */
+  TableIndices (const unsigned int index0,
+                const unsigned int index1);
+
+  /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<3> and initializes the
+   * indices stored by this object by the given arguments.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from three.
+   */
+  TableIndices (const unsigned int index0,
+                const unsigned int index1,
+                const unsigned int index2);
+
+  /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<4> and initializes the
+   * indices stored by this object by the given arguments.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from four.
+   */
+  TableIndices (const unsigned int index0,
+                const unsigned int index1,
+                const unsigned int index2,
+                const unsigned int index3);
+
+  /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<5> and initializes the
+   * indices stored by this object by the given arguments.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from five.
+   */
+  TableIndices (const unsigned int index0,
+                const unsigned int index1,
+                const unsigned int index2,
+                const unsigned int index3,
+                const unsigned int index4);
```

Comment:
Maybe something of this sort:
```
emplate <int N>
struct TableIndices
{
  template <typename... T>
  TableIndices (T... args)
    {
      static_assert (sizeof...(T) == N,
                     "Wrong number of constructor arguments!");
      values = {args...};
    }
private:
  unsigned int values[N];
};
```

The problem is that calling this as `TableIndices<1>(1.1)` (i.e., with a non-`unsigned int` argument yields the awkward error message
```
x.cc: In constructor ‘TableIndices<N>::TableIndices(T ...) [with T = double, int N = 1]’:
x.cc:15:   instantiated from here
x.cc:9: error: narrowing conversion of ‘args#0’ from ‘double’ to ‘unsigned int’ inside { }
x.cc: In constructor ‘TableIndices<N>::TableIndices(T ...) [with T = double, double, int N = 2]’:
```
I.e., you get the error message from inside the constructor, not from the place in user code where you use the wrong type. I'm not sure this is really important, given that these objects aren't typically generated by users, but I find this unsatisfying. It would be nice if there was a way to force the data types of the constructor to be of a type that *we* know, not as specified in user code...